src/mds/Locker.cc

```diff
@@ -225,6 +225,37 @@ struct MarkEventOnDestruct {
   }
 };
 
+/*
+ * Try to detect the possible deadlock between requests, and
+ * will try to break the newer request.
+ */
+void Locker::detect_dead_locks(const MDRequestRef& mdr, int lock_type)
+{
+  bool any_waiter = false;
+  bool order_mess = false;
+
+  dout(15) << "detect_dead_locks " << *mdr << dendl;
+  for (auto it = mdr->locks.begin(); it != mdr->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+
+    if (lock->is_waiter_for(SimpleLock::WAIT_ALL)) {
+      any_waiter = true;
+    }
+    if (lock->get_type() > lock_type) {
+      dout(1) << " detected mess lock order, locks: " << mdr->locks << dendl;
+      order_mess = true;
+    }
+    dout(15) << " lock " << *lock << " any_waiter " << any_waiter << " order_mess " << order_mess << dendl;
+
+    if (order_mess && any_waiter) {
```

Comment:
This needs to be turned on via a config. I think it should be off by default.

QA should turn it on and this method should message the clog with an ERR that we are resolving a possible deadlock so it fails tests.

Reply:
This sounds reasonable and I will add it.